code/modules/mob/living/carbon/xenomorph/mutation_datum.dm

```diff
@@ -0,0 +1,167 @@
+GLOBAL_DATUM_INIT(mutation_selector, /datum/mutation_datum, new)
+
+/datum/mutation_datum
+	interaction_flags = INTERACT_UI_INTERACT
+
+/datum/mutation_datum/ui_state(mob/user)
+	return GLOB.hive_ui_state // Similar purpose.
+
+/datum/mutation_datum/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "MutationSelector", "Mutation Selector")
+		ui.open()
+
+/datum/mutation_datum/ui_data(mob/living/carbon/xenomorph/xeno_user)
+	var/list/data = list()
+
+	data["shell_chambers"] = length(xeno_user.hive?.shell_chambers)
+	data["spur_chambers"] = length(xeno_user.hive?.spur_chambers)
+	data["veil_chambers"] = length(xeno_user.hive?.veil_chambers)
+	data["biomass"] = xeno_user.biomass
+	switch(xeno_user.xeno_caste.tier)
+		if(XENO_TIER_ONE)
+			data["cost"] = XENO_UPGRADE_BIOMASS_COST_T1
+		if(XENO_TIER_TWO)
+			data["cost"] = XENO_UPGRADE_BIOMASS_COST_T2
+		if(XENO_TIER_THREE)
+			data["cost"] = XENO_UPGRADE_BIOMASS_COST_T3
+		else
+			data["cost"] = XENO_UPGRADE_BIOMASS_COST_T4
+
+	var/list/upgrades = list(
```

Comment:
It'd be best that UI data is generated from the upgrade datums, instead of hardcoding ability names and their descriptions